components/core/src/clp/NetworkReader.cpp

```diff
@@ -0,0 +1,351 @@
+    return static_cast<NetworkReader*>(reader_ptr)->buffer_downloaded_data({ptr, size * nmemb});
+}
+}  // namespace
+
+bool NetworkReader::m_static_init_complete{false};
+
+auto NetworkReader::init() -> ErrorCode {
+    if (m_static_init_complete) {
+        return ErrorCode_Success;
+    }
+    if (0 != curl_global_init(CURL_GLOBAL_ALL)) {
+        return ErrorCode_Failure;
+    }
+    m_static_init_complete = true;
+    return ErrorCode_Success;
+}
+
+auto NetworkReader::deinit() -> void {
+#if defined(__APPLE__)
+    // NOTE: On macOS, calling `curl_global_init` after `curl_global_cleanup` will fail with
+    // CURLE_SSL_CONNECT_ERROR. Thus, for now, we skip `deinit` on macOS. Related issues:
+    // - https://github.com/curl/curl/issues/12525
+    // - https://github.com/curl/curl/issues/13805
+    // TODO: Remove this conditional logic when the issues are resolved.
+    return;
+#else
+    curl_global_cleanup();
+    m_static_init_complete = false;
+#endif
+}
+
+NetworkReader::NetworkReader(
+        std::string_view src_url,
+        size_t offset,
+        bool disable_caching,
+        std::chrono::seconds overall_timeout_int_sec,
+        std::chrono::seconds connection_timeout_in_sec,
+        size_t buffer_pool_size,
+        size_t buffer_size
+)
+        : m_src_url{src_url},
+          m_offset{offset},
+          m_file_pos{offset},
+          m_overall_timeout{overall_timeout_int_sec},
+          m_connection_timeout{connection_timeout_in_sec},
+          m_buffer_pool_size{std::max(cMinBufferPoolSize, buffer_pool_size)},
+          m_buffer_size{std::max(cMinBufferSize, buffer_size)} {
+    for (size_t i = 0; i < m_buffer_pool_size; ++i) {
+        // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays)
+        m_buffer_pool.emplace_back(std::make_unique<char[]>(m_buffer_size));
+    }
+    if (false == m_static_init_complete) {
+        throw OperationFailed(ErrorCode_NotReady, __FILE__, __LINE__);
+    }
+    m_downloader_thread = std::make_unique<DownloaderThread>(*this, offset, disable_caching);
+    m_downloader_thread->start();
+}
+
+NetworkReader::~NetworkReader() {
+    if (is_download_in_progress()) {
+        // Abort the download so the downloader thread can destroy the CURL resources and exit.
+        submit_abort_download_request();
+    }
+
+    while (is_downloader_thread_running()) {
+        // We could use sleep instead of busy waiting
+    }
+}
+
+auto NetworkReader::try_get_pos(size_t& pos) -> ErrorCode {
+    if (0 != m_offset) {
+        if (State::InProgress != get_state_code()) {
+            auto const curl_return_code{get_curl_return_code()};
+            if (false == curl_return_code.has_value()) {
+                return ErrorCode_Failure;
+            }
+            if (CURLE_HTTP_RETURNED_ERROR == curl_return_code.value()) {
+                // Download has failed due to http error. This can be caused by an out-of-bound
+                // offset specified in the http header.
```

Comment:
```suggestion
                // Download has failed due to an HTTP error. This can be caused by an out-of-bound
                // offset specified in the HTTP header.
```